File: GauravSir/Queue/QueueArray.c
#include<stdio.h>
#include<stdlib.h>
#include<assert.h>

typedef struct{
	int *data;
	int front;
	int rear;
	int size;
	int capacity;
}Queue;

Queue* createQueue(int capacity){
	assert(capacity>0);
	Queue *q=(Queue*)malloc(sizeof(Queue));
	assert(q!=NULL);
	q->data=(int*)malloc(capacity *sizeof(int));
	assert(q->data !=NULL);
	q->front=0;
	q->size=0;
	q->rear=-1;
	q->capacity=capacity;
	return q;
}

void resize(Queue *q,int cap){
	assert(cap>=q->size);
	int *newQ=(int*)malloc(sizeof(int)*cap);
	assert(newQ!=NULL);

	for(int i=0;i<q->size;i++){
		newQ[i]=q->data[(q->front+i)% q->capacity];
	}

	free(q->data);
	q->data=newQ;
	q->front=0;
	q->rear=q->size-1;
	q->capacity=cap;
}

void enqueue(Queue*q,int val){
	if(q->size==q->capacity){
		resize(q,q->capacity*2);
	}
	q->rear=(q->rear+1)%q->capacity;
	q->data[q->rear]=val;
	q->size++;
}

int dequeue(Queue*q){
	if(q->size==0)return -1;

	int val=q->data[q->front];
	q->front=(q->front+1)%q->capacity;
	q->size--;

	if(q->size<q->capacity/2&&q->capacity>4){
		int newC=(q->capacity*3)/4;
		resize(q,newC);
	}
	return val;
}

void print(Queue*q){
	for(int i=0;i<q->size;i++){
		printf("%d ",q->data[(q->front+i)%q->capacity]);
		
	}

	printf("\n");
}

void freeQueue(Queue*q){
	free(q->data);
	free(q);
}

int main(){
	Queue *q=createQueue(4);

	enqueue(q,10);
	enqueue(q,20);
	enqueue(q,30);
	enqueue(q,40);
	enqueue(q,50);
	enqueue(q,60);

	print(q);
	dequeue(q);
	dequeue(q);
	print(q);
	freeQueue(q);

}

File: GauravSir/Queue/QueueArray.java
class Queue {
    private int[] data;
    private int front;
    private int rear;
    private int size;
    private int capacity;

    public Queue(int capacity) {
        if (capacity <= 0)
            throw new IllegalArgumentException("Capacity must be greater than 0");
        this.capacity = capacity;
        this.data = new int[capacity];
        this.front = 0;
        this.rear = -1;
        this.size = 0;
    }

    private void resize(int newCapacity) {
        if (newCapacity < size)
            throw new IllegalArgumentException("New capacity must be >= current size");

        int[] newData = new int[newCapacity];
        for (int i = 0; i < size; i++) {
            newData[i] = data[(front + i) % capacity];
        }

        data = newData;
        front = 0;
        rear = size - 1;
        capacity = newCapacity;
    }

    public void enqueue(int val) {
        if (size == capacity) {
            resize(capacity * 2); // double capacity when full
        }

        rear = (rear + 1) % capacity;
        data[rear] = val;
        size++;
    }

    public int dequeue() {
        if (size == 0) {
            System.out.println("Queue underflow");
            return -1;
        }

        int val = data[front];
        front = (front + 1) % capacity;
        size--;

        if (size < capacity / 2 && capacity > 4) {
            int newCap = (capacity * 3) / 4;
            resize(newCap);
        }

        return val;
    }

    public void print() {
        for (int i = 0; i < size; i++) {
            System.out.print(data[(front + i) % capacity] + " ");
        }
        System.out.println();
    }

    public void freeQueue() {
        data = null; // in Java, GC will reclaim automatically
    }

    public boolean isEmpty() {
        return size == 0;
    }
}

public class QueueArray {
    public static void main(String[] args) {
        Queue q = new Queue(4);

        q.enqueue(10);
        q.enqueue(20);
        q.enqueue(30);
        q.enqueue(40);
        q.enqueue(50);
        q.enqueue(60);

        q.print();

        q.dequeue();
        q.dequeue();
        q.print();

        q.freeQueue(); // optional in Java
    }
}
